pkg/compact/compact.go

```diff
@@ -470,6 +481,102 @@ func (cg *Group) Resolution() int64 {
+type PlanSim interface {
+	ProgressCalculate(ctx context.Context, groups []*Group) error
+}
+
+type DefaultPlanSim struct {
+	planner Planner
+	*ProgressMetrics
+}
+
+func NewDefaultPlanSim(reg prometheus.Registerer, planner Planner) *DefaultPlanSim {
+	return &DefaultPlanSim{
+		planner: planner,
+		ProgressMetrics: &ProgressMetrics{
+			NumberOfIterations: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+				Name: "thanos_number_of_iterations",
+				Help: "number of iterations to be done",
+			}, []string{"group"}),
+			NumberOfBlocksToMerge: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+				Name: "thanos_number_of_blocks_planned",
+				Help: "number of blocks planned to be merged",
+			}, []string{"group"}),
+		},
+	}
+}
+
+func (ps *DefaultPlanSim) ProgressCalculate(ctx context.Context, groups []*Group) error {
+	iterations := 0
+	blocksToMerge := 0
+	groupCompactions := make(map[string]int, len(groups))
+	groupBlocks := make(map[string]int, len(groups))
+
+	hasPlan := true
+	for hasPlan {
+		for _, g := range groups {
+			if len(g.IDs()) == 1 {
+				continue
+			}
+
+			// parameter should be of type tsdb.BlockMeta.meta
+			plan, err := ps.planner.Plan(ctx, g.metasByMinTime)
+			if err != nil {
+				return errors.Wrapf(err, "could not plan")
+			}
+			if len(plan) == 0 {
+				continue
+			}
+			iterations++
+			groupCompactions[g.key] = iterations
+
+			// value type in map is struct{} - enpty struct consumes 0 bytes so prefered over bool
+			toRemove := make(map[ulid.ULID]struct{}, len(plan))
+			metas := make([]*tsdb.BlockMeta, 0, len(plan))
+			for _, p := range plan {
+				metas = append(metas, &p.BlockMeta)
+				toRemove[p.BlockMeta.ULID] = struct{}{}
+			}
+			g.deleteFromGroup(toRemove)
+
+			blocksToMerge += len(plan)
+			groupBlocks[g.key] = blocksToMerge
+
+			// remove 'plan' blocks from 'original metadata' - so that the remaining blocks can now be planned ?
+			// not required to modify originalMeta now
+
+			if len(g.metasByMinTime) == 0 {
+				hasPlan = false
+				break
+				// no plan in case the group is empty after removing the 'planned' metadata
+				// group size will remain one even after newMeta is added, hence, no plan needed for this case
+			}
+
+			newMeta := tsdb.CompactBlockMetas(ulid.MustNew(uint64(time.Now().Unix()), nil), metas...)
+			g.AppendMeta(&metadata.Meta{BlockMeta: *newMeta})
+		}
+	}
+
+	// updated only once here - after the entire planning simulation is completed
+	// updating the exposed metrics inside the above loop will change based on iterations needed for each plan loop
+	// updating the metrics' maps directly - some keys may not be present in the groups map; also saves the cost of a lookup if done directly
+	for key, iters := range groupCompactions {
+		ps.ProgressMetrics.NumberOfIterations.WithLabelValues(key).Add(float64(iters))
+	}
+	for key, blocks := range groupBlocks {
+		ps.ProgressMetrics.NumberOfBlocksToMerge.WithLabelValues(key).Add(float64(blocks))
```

Comment:
You can add this line to the previous for loop.
If there is a compaction iteration, then this group must also have blocks to merge. No need to iterate the map twice.